Check for overflow in DroplessArena and align returned memory

tmiasko opened the pull request:
* Check for overflow when calculating the slice start & end position.
* Align the pointer obtained from the allocator, ensuring that it
  satisfies user requested alignment (the allocator is only asked for
  layout compatible with u8 slice).
* Remove an incorrect assertion from DroplessArena::align.
* Avoid forming references to an uninitialized memory in DroplessArena.

Helps with #73007, #72624.

tmiasko committed:
Check for overflow in DroplessArena and return aligned pointer

* Check for overflow when calculating the slice start & end position.
* Align the pointer obtained from the allocator, ensuring that it
  satisfies user requested alignment (the allocator is only asked for
  layout compatible with u8 slice).
* Remove an incorrect assertion from DroplessArena::align.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @petrochenkov (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

petrochenkov commented:
r? @nnethercote

nnethercote commented on src/librustc_arena/lib.rs:
```diff
@@ -371,21 +364,27 @@ impl DroplessArena {
-            assert!(bytes != 0);
-
-            self.align(align);
+    fn alloc_raw_fast(&self, bytes: usize, align: usize) -> Option<&mut [u8]> {
```
Would `_without_growth` be better than `_fast`?

nnethercote commented on src/librustc_arena/lib.rs:
```diff
@@ -371,21 +364,27 @@ impl DroplessArena {
+        loop {
+            if let Some(a) = self.alloc_raw_fast(bytes, align) {
+                break a;
             }
```
What happens here on overflow? It'll call `self.grow(bytes)`. Then what will happen -- panic, success? I think the possibilities need to be clearer here. In particular, `alloc_raw_fast()` returns `None` in two cases: "overflow", and "need to grow". I feel like those two cases need to be distinguished.

tmiasko replied:
From perspective of code that tried to use the current chunk the situation is the same: there is no capacity left and arena needs to grow. Overflow does not necessarily preclude later success if say the chunk had merely a high address, requested capacity was large, and the calculation overflowed.

The code then delegates the responsibility to handle the situation to `grow` and subsequently to `RawVec`. Under assumption that grow doesn't hit `checked_mul(2).unwrap()` (this could be saturated to `isize::MAX` to be a little bit friendlier to 32-bit systems), the allocation is performed by `RawVec` which panics when requested capacity exceeds `isize::MAX` (on 32-bit systems) and aborts on OOM.

nnethercote replied:
Thanks for the explanation. Could you put that information into a comment? :)

tmiasko replied:
Now with additional comments.

nnethercote replied:
Thanks for that. Sorry to nitpick this further, but you've added a comment explaining the high-level behaviour, which is fairly obvious, while not adding a comment distinguishing the overflow vs. need-to-grow cases, which is less obvious.

Can you add a comment inside the function explaining those two cases, much like your comment above? I specifically would like to see explanation of the "high address + large capacity" case, in particular. Thanks.

tmiasko replied:
I added a *different* comment explaining the abstract check we want to perform.
In my view the overflow is the same as the need-to-grow case, which is why I
didn't want to comment about the distinction. I just don't see any.

tmiasko committed:
Avoid forming references to an uninitialized memory in DroplessArena

Return a pointer from `alloc_raw` instead of a slice. There is no
practical use for slice as a return type and changing it to a pointer
avoids forming references to an uninitialized memory.

nnethercote commented on src/librustc_arena/lib.rs:
```diff
@@ -370,22 +363,33 @@ impl DroplessArena {
-            // Set the pointer past ourselves
-            self.ptr.set(intrinsics::arith_offset(self.ptr.get(), bytes as isize) as *mut u8);
-            slice::from_raw_parts_mut(ptr, bytes)
+            // No free space left. Allocate a new chunk to satisfiy the request.
```
"satisfy"

nnethercote commented:
@bors r+

bors commented:
:pushpin: Commit 1f0895162ba5a783d4d73d5c263552eaca9343b3 has been approved by `nnethercote`

<!-- @bors r=nnethercote 1f0895162ba5a783d4d73d5c263552eaca9343b3 -->
<!-- homu: {"type":"Approved","sha":"1f0895162ba5a783d4d73d5c263552eaca9343b3","approver":"nnethercote"} -->

yoshuawuyts commented:
@tmiasko I didn't realize #73007 was merged; thanks so much for enabling that with this PR! :pray:
